src/llvm/CompilerIntrinsics.js

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import {
+  Module,
+  Function as LLVMFunction,
+  FunctionType,
+  StructType,
+  Type as LLVMType,
+  LinkageTypes,
+  config,
+} from "llvm-node";
+import { llvmContext } from "./llvm-context.js";
+
+export class Intrinsics {
+  +_module: Module;
+  _stringType: ?StructType;
+  _uint32Type: ?StructType;
+  _memcpy: ?LLVMFunction;
+  _memcmp: ?LLVMFunction;
+  _pow: ?LLVMFunction;
+
+  constructor(module: Module) {
+    this._module = module;
+  }
+
+  isStringType(type: LLVMType): boolean {
+    if (!this._stringType) return false;
```

Comment:
Is it a good idea to return false if _stringType is not defined? Why not return this.stringType.equals(type)?

Reply:
If it's not defined, it'll throw and Flow won't let me call equals on it.

If it is not yet defined, that means that there are no strings defined in this context yet. So whatever this type is, it can't be equal to a string since there are no strings.

Reply:
Hmm, this is extremely subtle. You are essentially saying that the only way a there can be a type T so that this.stringType.equals(T) is if T === this.stringType. I'm none too sure that this is desirable, but if it is, there should be a big comment about it and my follow up to that would be why call equals instead of using ===?

Reply:
The reason for this approach is because I don't want to generate unnecessary code in the LLVM output for a bunch of types if this type is never referenced in the output. I still have to test whether something is this type without knowing if it might be though. I suspect there will be a lot more of these cases for other types that I'll need to encode. I wouldn't want to document it for ever type but I can make a global comment about it in this file.

The `equals` call is just how the LLVM node bindings work since the wrapper objects don't guarantee JS object equality since the underlying C++ APIs can return instances that the JS bindings doesn't know about. That pattern will be all over the code base. It's pretty common since JS doesn't have operator overloading (yet).

Reply:
I'm still befuddled. If the C++ API can return a value that wasn't obtained via the getter of stringType, what guarantee is there that isStringType won't be called before stringType is called and therefore have inconsistent behavior?

Reply:
The C++ can't make up a type that I haven't created yet. However, if I have created it, the C++ API can return the C++ version of that object from reflection APIs. The JS binding around the C++ bindings doesn't know that there already is a JS instance associated with this C++ object so it creates a copy.

Reply:
OK, be that as it may then. A source code comment to this effect would be very helpful to a future reader. Ideally some invariant to enforce it would also be desirable.